ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapUtils.java

```diff
@@ -324,6 +325,23 @@ static boolean areAllReplicasForPartitionUp(PartitionId partition) {
     return true;
   }
 
+  /**
+   * Check whether all local replicas of the given {@link PartitionId} are up.
+   * @param partition the {@link PartitionId} to check.
+   * @param localDatacenterName Name of the local datacenter
+   * @return true if all associated replicas are up; false otherwise.
+   */
+  static boolean areAllLocalReplicasForPartitionUp(PartitionId partition, String localDatacenterName) {
+    for (ReplicaId replica : partition.getReplicaIds()) {
+      if (replica.getDataNodeId().getDatacenterName().equals(localDatacenterName) && replica.isDown()) {
+        logger.debug("Replica [{}] on {} {} is down", replica.getPartitionId().toPathString(),
+            replica.getDataNodeId().getHostname(), replica.getMountPath());
+        return false;
+      }
```

Comment:
I think you can maintain a Map inside `PartitionSelectionHelper`, which will be populated during initialization. The Map keeps each partition and a list of its local replicas, i.e `Map<PartitionId, List<ReplicaId>>`. 
The benefit is, when this method is called, it only needs to check 3 replicas directly rather than 12 replicas in worst case.

Reply:
@ankagrawal, did you want to make this change or should I merge the PR?